source/agora/script/Engine.d

```diff
@@ -0,0 +1,1002 @@
+
+        Execute a P2SH (Pay 2 Script Hash) type of lock script.
+
+        The 64-byte hash of the redeem script `H` is read from `lock_bytes`,
+        `unlock_bytes` is evaluated as a set of pushes to the stack where
+        the last push is the redeem script. The redeem script is popped from the
+        stack, hashed, and compared to `H` from the lock script. Then it's
+        evaluated with any leftover stack items.
+
+        Params:
+            lock = must contain a 64-byte hash of the redeem script
+            unlock = must contain only stack push opcodes, where the last
+                           push is the redeem script itself
+            tx = the associated spending transaction
+
+        Returns:
+            null if there were no errors,
+            or a string explaining the reason execution failed
+
+    ***************************************************************************/
+
+    private string executeScriptHash (Lock lock, Unlock unlock,
+        in Transaction tx) nothrow @safe
+    {
+        assert(lock.type == LockType.Redeem);
+
+        if (lock.bytes.length != Hash.sizeof)
+            return "Lock script: LockType.Redeem requires 64-byte script hash argument";
+        const Hash script_hash = Hash(lock.bytes);
+
+        Script unlock_script = Script(unlock.bytes);
+        if (auto error = unlock_script.isInvalidSyntaxReason(ScriptType.Unlock))
+            return error;
+
+        Stack stack;
+        if (auto error = executeScript(unlock_script, stack, tx))
+            return error;
+
+        if (stack.empty())
+            return "Unlock script did not push a redeem script to the stack";
+
+        const redeem_bytes = stack.pop();
+        if (hashFull(redeem_bytes) != script_hash)
+            return "Hash of redeem script does not match redeem hash in lock script";
+
+        Script redeem = Script(redeem_bytes);
+        if (auto error = redeem.isInvalidSyntaxReason(ScriptType.Redeem))
+            return error;
+
+        if (auto error = executeScript(redeem, stack, tx))
+            return error;
+
+        if (hasStackFailed(stack))
+            return "Script failed";
+
+        return null;
+    }
+
+    /***************************************************************************
+
+        Execute the script with the given stack and the associated transaction
+
+        Params:
+            script = the script to execute
+            stack = the stack to use for the script. May be non-empty.
+            tx = the associated spending transaction
+
+        Returns:
+            null if there were no errors,
+            or a string explaining the reason execution failed
+
+    ***************************************************************************/
+
+    private string executeScript (in Script script, ref Stack stack,
+        in Transaction tx) nothrow @safe
+    {
+        ScopeCondition sc;
+        const(ubyte)[] bytes = script[];
```

Comment:
It seems a bit strange that we have a `Script` that is built using `ubyte[]`, then manually validated, and then in here it's sliced again.

It would be better to move the syntactical validation into a static routine which **returns** a valid script. That way we can know for sure that when a function receives a `Script` parameter, it is by definition valid.